### PR TITLE
applications: nrf_desktop: Add buttons documentation

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:DESK-667

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by MarekPieta** [comment]
We may need to write something about scanning limitations and anti-ghosting

File: applications/nrf_desktop/doc/buttons.rst [hunk 1]
```diff
@@ -1,7 +1,65 @@
...
+In ``STATE_ACTIVE`` module enables GPIO interrupts and wait for pins state to
+change. When any button is pressed module switches to ``STATE_SCANNING``.
+When switch occurs module submits a work with a delay set to
+``CONFIG_DESKTOP_BUTTONS_DEBOUNCE_INTERVAL``. Work performs scanning of the
+keyboard matrix noting changes to button states and sending related events.
+While any button is kept pressed while scanning is performed work will be
+re-submitted with delay set to ``CONFIG_DESKTOP_BUTTONS_SCAN_INTERVAL``.
+If no button is pressed module switch back to ``STATE_ACTIVE``.
+
+When system enters low-power state, ``buttons`` goes to ``STATE_IDLE`` in which
```
Comment (greg-fer) [inline applications/nrf_desktop/doc/buttons.rst]:
```suggestion
When the system enters the low-power state, the ``buttons`` module goes to ``STATE_IDLE``, in which
```

File: applications/nrf_desktop/doc/buttons.rst [hunk 2]
```diff
@@ -1,7 +1,65 @@
...
+it waits for GPIO interrupts indicating change to button states. When interrupt
```
Comment (greg-fer) [inline applications/nrf_desktop/doc/buttons.rst]:
```suggestion
it waits for GPIO interrupts that indicate a change to button states. When an interrupt
```

File: applications/nrf_desktop/doc/buttons.rst [hunk 3]
```diff
@@ -1,7 +1,65 @@
...
+is triggered module will issue a system wake up event.
```
Comment (greg-fer) [inline applications/nrf_desktop/doc/buttons.rst]:
```suggestion
is triggered, the module will issue a system wakeup event.
```

File: applications/nrf_desktop/doc/buttons.rst [hunk 4]
```diff
@@ -1,7 +1,65 @@
...
+In case request for power off comes while button is held, module switches to
```
Comment (greg-fer) [inline applications/nrf_desktop/doc/buttons.rst]:
```suggestion

If a request for power off comes while the button is pressed, the module switches to
```
Move this paragraph to a new paragraph after line 59.

File: applications/nrf_desktop/doc/buttons.rst [hunk 5]
```diff
@@ -1,7 +1,65 @@
...
+``STATE_SUSPENDING`` and be in this state until no button is pressed. Then it
```
Comment (greg-fer) [inline applications/nrf_desktop/doc/buttons.rst]:
```suggestion
``STATE_SUSPENDING`` and remains in this state until no button is pressed. Then, it
```

File: applications/nrf_desktop/doc/buttons.rst [hunk 6]
```diff
@@ -1,7 +1,65 @@
...
+
+
+Configuration
+*************
+
+Module is enabled by ``CONFIG_DESKTOP_BUTTONS_ENABLE``.
+
+Module can handle both matrix keyboard and buttons connected directly to GPIO pins.
+
+When defining how button are connected two arrays are used. First one contains
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/doc/buttons.rst]: fixed

File: applications/nrf_desktop/doc/buttons.rst [hunk 7]
```diff
@@ -1,7 +1,65 @@
...
+
+By default it is assumed that button press is indicated by pin switch from low
+to high state. This can be changed with ``CONFIG_DESKTOP_BUTTONS_POLARITY_INVERSED``
+that will cause application react to pin change from high to low state.
+
+Implementation details
+**********************
+
+``buttons`` can be in the following states:
+
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/doc/buttons.rst]: fixed

File: applications/nrf_desktop/doc/buttons.rst [hunk 8]
```diff
@@ -1,7 +1,65 @@
...
+    * STATE_IDLE
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/doc/buttons.rst]: done

File: applications/nrf_desktop/doc/buttons.rst [hunk 9]
```diff
@@ -1,7 +1,65 @@
...
+    * STATE_ACTIVE
+    * STATE_SCANNING
+    * STATE_SUSPENDING
+
+In ``STATE_ACTIVE`` module enables GPIO interrupts and wait for pins state to
+change. When any button is pressed module switches to ``STATE_SCANNING``.
+When switch occurs module submits a work with a delay set to
+``CONFIG_DESKTOP_BUTTONS_DEBOUNCE_INTERVAL``. Work performs scanning of the
+keyboard matrix noting changes to button states and sending related events.
+While any button is kept pressed while scanning is performed work will be
```
[the remark this answers is not in the document]
Comment (pdunaj) [reply applications/nrf_desktop/doc/buttons.rst]: Yes, but scanning will continue while any button is pressed. E.g. we press A -> scanning is started and performed every T time. If A is released but B is pressed between scans, scanning will continue. If there is no button pressed while scanning, scan will stop and module will switch to an interrupt based mode.

**Comment by pdunaj** [comment]
@wiminordic , @jhn-nordic could you give me a green tick?

@carlescufi , would it be possible for @greg-fer to have a green tick?

**Comment by carlescufi** [comment]
> @wiminordic , @jhn-nordic could you give me a green tick?
> 
> @carlescufi , would it be possible for @greg-fer to have a green tick?

@pdunaj  sorry for the delay.

@greg-fer does not have that access, but I can give it to him if we think it's appropriate.

**Pull request merged**